server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -112,7 +115,25 @@ private static XContentBuilder mappings() {
         }
     }
 
-    public void getSynonymsSet(String resourceName, int from, int size, ActionListener<SynonymsSetResult> listener) {
+    public void getSynonymsSets(int from, int size, ActionListener<PagedResult<SynonymSetSummary>> listener) {
+        client.prepareSearch(SYNONYMS_ALIAS_NAME)
+            .setSize(0)
+            .addAggregation(new TermsAggregationBuilder(SYNONYM_SETS_AGG_NAME).field(SYNONYMS_SET_FIELD).order(BucketOrder.key(true)))
+            .setPreference(Preference.LOCAL.type())
+            .execute(listener.delegateFailure((searchResponseListener, searchResponse) -> {
+                Terms aggregation = searchResponse.getAggregations().get(SYNONYM_SETS_AGG_NAME);
+                List<? extends Terms.Bucket> buckets = aggregation.getBuckets();
+                SynonymSetSummary[] synonymSetSummaries = buckets.stream()
+                    .skip(from)
```

Comment:
I am not sure if we need `from` parameter at all. I assume there will be a very few synonyms_set.

Reply:
Can we try to not rely on aggregations? I am not sure why we're using them but it feels odd that we need them on a system index. We should try to rely on get and a very simple search that works only on keyword exact match, aggregations should be out of scope and should probably be disabled on system indices in general.

Reply:
> I am not sure if we need from parameter at all

We had mentioned the need for [using pagination on this API](https://docs.google.com/document/d/1ZIJER4qEab-OacZuPMYRP57g-T45ZkuCvqOnJSCBlUs/edit#heading=h.61r44hllliq4). We can remove that restriction as we probably won't have thousands of Synonym Sets. WDYT?

> Can we try to not rely on aggregations? 

We're trying to get the different synonym sets that exist (the different values for a field), so I assumed that a term aggregation would be the better way to go. We need to think that there could be thousands of synonym rules, so collecting them on an iterative basis seemed wasteful to me.

We could create a different document type (or a different index) to store just the synonym set information, and use the current structure for synonym rules. WDYT?

Reply:
I'm starting to like the idea of using a different doc type / index to store the synonym sets. That would allow us to:

- Store empty synonym sets. Right now, a synonym set with all its rules removed does not exist.
- Provide sync mechanisms. There's the possibility of race conditions that could be solved using version control on the synonym set documents.

We could of course revisit the decision of using a single doc for storing a synonym set. I still struggle with pagination and searching though, but that's probably doable from the UI side holding all data in memory. 🤔 🔥

Reply:
I think we should focus on what we can do with the current model. We spent time on choosing the right one so now we should also acknowledge the limitations of our decision. In that spirit I'd say that listing all the synonym set should be removed from our requirements. We can have a link on the index if it has a reference to a synonym collection and that would be enough. We can revisit this later so let's not spend too much time on this.

Reply:
I think listing the synonym set is important, otherwise if a user forgets the name of a synonym set they put, they can't do anything.  

> have a link on the index if it has a reference to a synonym collection and that would be enough
I am not sure what you mean by this, can you please elaborate on this.

----

@jimczi  @carlosdelest As an alternative to a term aggregation, we can use [terms enum API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-terms-enum.html). It should be super fast, as it lists only terms. But the API will change: we will not have `count` in response, and we don't need request params of `size` or `from`. What do you think?

Reply:
I agree with Mayya that we need to provide a way to list the synonym sets - that would be mandatory from a discovery perspective.

I like the Terms API idea, thanks Mayya! We could still use size and from, but given your points on pagination, I might remove that entirely and simplify the design. I'll update the code with the Terms API and we can keep the discussion 👍

Reply:
Got it, and it looks like Terms enum API also support pagination.

Reply:
Sorry, it turns out we can't it use `terms_enum` as that code is in X-Pack. How about using `wildcard` query?

Reply:
Another ideas is to use _meta field to store synonyms sets, but I have not thought much of it, we should explore it.

Reply:
We can also move `terms_enum` to core, it doesn't need to be in x-pack per say.

Reply:
There seems to be some [intermediate work to be done](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java#L388) for moving `terms_enum` to core, as some of the work that needs to be done from TransportTermsEnumAction has to deal with Security, and thus access the XPack related classes.

Could we open a separate task for doing this, and in the meantime use aggregations?

Another option would be to use a separate index or document types to store the synonym sets. That would allow us to do optimistic concurrency control on synonym sets using these documents. The downsides would be:
- More complex design
- In order to obtain the number of synonym rules for each synonym set, the UI or clients would need to query each synonym set individually instead

Reply:
My bad, at least we tried ;). Let's use aggregations for now, if we're careful the aggregation should be cheap since it will use filters on a low cardinality field.